## Importing Libraries

In [320]:
##Importing important packages

import pandas as pd
import numpy as np
import os


In [321]:
##Importing path, and csv files of products

path= r'/Users/erictam/Downloads/CareerFoundry Program/Data Immersion Achievement 4/Instacart Basket Analysis'

# To import products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', '4.3 products.csv'), index_col=False)

# To import orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', '4.4 orders_wrangled.csv'), index_col=False)


In [322]:
#Creating a dataframe

df_test= pd.DataFrame()

In [323]:
#Creating a mixed type column

df_test['mix']=['a','b',1,True]

In [324]:
#Checking for mixed-types

for col in df_prods.columns.tolist():
    weird = (df_prods[col].map(type) != type(df_prods[col].iloc[0])).any()
    if weird:
        print(col)

product_id
product_name
aisle_id
department_id
prices


In [325]:
#Function to count any missing values 

df_prods.isnull().sum()


product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [326]:
#Subsetting df_prods for only the 16 missing product_name

df_nan =  df_prods[df_prods['product_name'].isnull()==True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [327]:
#Subsetting df_prods_clean to remove missing values by setting .isnull() to false

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]


In [328]:
#Finding duplicate rows of the df_prods_clean dataframe

df_dups = df_prods_clean[df_prods_clean.duplicated()]


#Making new df using drop_duplicates function
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()
df_prods_clean_no_dups

#This new Df_prods_clean_no_dups is exactly 5 less than the df_prods_clean so we know it was done correctly
#Will rename this into "products_checked.csv" when exporting since df_prods_clean_no_dups can be long

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

## 4.5 Task Section Header

In [330]:
# 2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

df_ords.describe()



Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_time_hour_of_day  days_since_prior_order  
count        3.421083e+06            3.421083e+06            3.214874e+06  
mean         2.776219e+00            1.345202e+01            1.111484e+01  
std          2.046829e+00            4.226088e+00            9.206737e+00  
min          0.000000e+00            0.000000e+00            0.000000e+00  
25%          1.000000e+00            1.000000e+01            4.000000e+00  
50%          3.000000e+00            1.300000e+01            7.000000e+00  
75%          5.000000e+00            1.600000e+01            1.500000e+01  
max          6.000000e+00            2.300000e+01            3.000000e+01


orders_day_of_week: The min and max are from 0 to 6, which represent Sunday through Saturday. These should all be whole numbers and is reforced by the 50% mid point being exactly 3. This means that there are no float variables that may skew the data

order_time_hour_of_day: The min and max times are 0 and 23. This is in military time and it make sense that there is not a 24 as that is equivalent to 0 for time. 

days_since_prior_order: All of these values are (+) values, along with the other columns as there cannot be a negative time. 

In [332]:
# 3. Check for mixed-type data in your df_ords dataframe.


# This code was copied from the CareerFoundry website. It continued to receive an error message that the DataFrame.applymap is deprecate and possibly going to be removed upon futuer updates. Online research said to use map instead of applymap to avoid this
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)     



/var/folders/c3/gxzbvq9x1032lnrpvbyhdtrw0000gn/T/ipykernel_24438/3033560794.py:6: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/c3/gxzbvq9x1032lnrpvbyhdtrw0000gn/T/ipykernel_24438/3033560794.py:6: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/c3/gxzbvq9x1032lnrpvbyhdtrw0000gn/T/ipykernel_24438/3033560794.py:6: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/c3/gxzbvq9x1032lnrpvbyhdtrw0000gn/T/ipykernel_24438/3033560794.py:6: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].ilo

In [333]:
# 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

#There was no mixed types found within the column. However, there are many nulls when using the isnull() function on df_ords

#5. Run a check for missing values in your df_ords dataframe.
#In a markdown cell, report your findings and propose an explanation for any missing values you find.

    
df_ords.isnull().sum()



Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_time_hour_of_day         0
days_since_prior_order    206209
dtype: int64

5. 
isnull() finds 206209 null values in this column. This data is an integer and imputation can be applicable. We will imputate using the mean value.

In [335]:
#6. Address the missing values using an appropriate method

df_ords['days_since_prior_order'] = df_ords['days_since_prior_order'].fillna(df_ords['days_since_prior_order'].mean())
df_ords



Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_time_hour_of_day  days_since_prior_order  
0                             8               11.114836  
1                             7               15.000000  
2                            12               21.000000  
3                             7               29.000000  
4                            15               28.000000  
...                         ...                     ...  
3421078                      18               29.000000  
3421079                      10               30.000000  
3421080                      12               18.000000  
3421081                      12                7.000000  
3421082                      14               30.000000  

[3421083 rows x 7 columns]

6. 
This data in days_since_prior_order is an integer and imputation can be applicable. We will imputate using the mean value since it is provides a middle average of the days data.


In [359]:
# 7.Run a check for duplicate values in your df_ords data.
# In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

# 8. Address the duplicates using an appropriate method.
# In a markdown cell, explain why you used your method of choice.

    
df_ords_dups = df_ords[df_ords.duplicated()]
df_ords_dups.head()

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_time_hour_of_day, days_since_prior_order]
Index: []

7 and 8. From the duplicated() function, it shows that there are no duplicated rows, so no further action needs to be done. If there were, then we would go through process of deleting them through the drop_duplicates() function

## Exporting new csv files

In [368]:


#Exporting df_ords
df_ords.to_csv(os.path.join(path,'02 Data','Prepared Data','4.5 orders_checked.csv'))


#Exporting df_dep_t_new

df_prods.to_csv(os.path.join(path,'02 Data','Prepared Data','4.5 prods_checked.csv'))
